osu.Game/Rulesets/Objects/Drawables/DrawableHitObject.cs

```diff
@@ -687,12 +687,14 @@ public virtual void OnKilled()
         /// the <see cref="ScoreProcessor"/> of the <see cref="JudgementResult"/>.
         /// </summary>
         /// <param name="application">The callback that applies changes to the <see cref="JudgementResult"/>.</param>
-        protected void ApplyResult(Action<JudgementResult> application)
+        /// <param name="state">The state passed to the callback.</param>
+        /// <typeparam name="TState">The type of the state information that is passed to the callback method.</typeparam>
+        protected void ApplyResult<TState>(Action<JudgementResult, TState> application, TState state)
```

Comment:
I would like to hope that this sort of construction - passing arbitrary `TState` - can be avoided. It's not great API design - as evidenced by the fact that the description of the typeparam reads like opaque word salad. The xmldoc should either explain _what_ `TState` is for (reducing closure allocation), _or_, better yet, the generic should be gone and the closure should just accept `this`. As in the signature should be

```csharp
protected void ApplyResult(Action<JudgementResult, DrawableHitObject> application);
```

Seems like this should be trivially achievable (and is already in fact done in most usages of this contraption).

Reply:
Using `TState` is necessary to fully avoid the closure allocations in all cases. For example, `this` is not sufficient enough in what id assume to be the hottest path `DrawableHitCircle.CheckForResult`. If there is an opportunity to completely rid of the allocation in all cases by use of `TState` then I think that is the solution that should be used. I think it's pretty simple and flexible. Both the use and implementation are straightforward. If the only problem with the API is the typeparam reading poorly then I'm all for changing the wording in the xmldoc. I agree it can be worded better. If you agree with what I've said here, I will make changes to the xmldoc once we agree on better language. 

English isn't my strong suite, I stopped showing up senior year of high school and up until then I'd fall asleep in class. Here is alternative wording I used GPT to help me write:

```
/// <param name="state">The state passed to the callback. This parameter is used to pass additional data or context 
/// required by the callback method. It is specifically designed to reduce closure allocations by allowing state 
/// information to be passed explicitly, instead of allocating a closure.</param>
/// <typeparam name="TState">The type of the state information passed to the callback. This generic type 
/// parameter allows for passing a variety of data structures, enabling flexibility while minimizing memory 
/// allocations associated with closures.</typeparam>
```

If you like this language more I will update the code.

Reply:
> For example, `this` is not sufficient enough in what id assume to be the hottest path `DrawableHitCircle.CheckForResult`.

I dunno, is it? If it's the `ResultFor()` call that you're hinting at, then I see at least two options of eliminating it:

- `ResultFor()` is pure, so you could just as well call it again inside the callback (requires nothing more than a `this` capture)
- Or you could just store the result to a field (also requires nothing more than a `this` capture).

> Here is alternative wording I used GPT to help me write:

Gonna be honest I do not look upon that favourably. I'd call GPT's output there "pure content", in that it's style over substance. I'm not asking for excellence in English, I'm asking for something along the lines of

```csharp
/// <param name="state">
/// Use this parameter to pass any data that <paramref name="application"/> requires
/// to apply a result. Doing things this way avoids closure allocations.
/// </param>
/// <typeparam name="TState">The type of <paramref name="state">.</typeparam>
```

Reply:
Thank you for the feedback on the xmldoc language, I understand where you are coming from better now. I like what you've wrote. Its a better, more concise, version of what I was originally aiming for. I was a little confused originally by your xmldoc feedback and that's why I went to GPT to try and help me write something more full.

I agree it's probably best if we get others involved in the conversation. The `ApplyResult` API is also used by user implemented rulesets. I'm not familiar with all the data accessible from within each of these classes so I'm not fully able to assess if using `this` in all cases can result in all the closures being avoided. From my surface level judgement it looks like multiple class definitions will need added fields in order to avoid the closure allocations. I also do not know if others want to change the class definitions to contain more fields. I personally wouldn't want to change the class definitions, that's just my opinion, but maybe there is a benefit to doing so I'm unaware of. I think whats important is making sure these closure allocations are avoided. The faster the map, the more hit circles it has, the more frequently these allocations will occur. There is an opportunity to avoid them in full so I think the opportunity should be taken.

Reply:
Are you just a GPT bot? Your replies are 100% gpt. Please don't.

Reply:
Accepting `this` should be the only path forward here.

Reply:
lmao I promise you I did not use GPT to write anything other than that bit of xmldoc, I was just confused on what "opaque word salad" meant so that's why I turned to GPT to try and help me understand. Accepting `this` requires new fields get added to classes to fully avoid closure allocations across all cases.

Reply:
@bdach can you have a look at the diff and see if you're okay with the fields in `DrawableHitCircle`? I have some improvements to this PR which *do* involve having a `TState` method, but the only `protected` usage of this is `DrawableHitCircle` for now.

I want to know if you prefer it or not. If not, see what you think of my other changes and if they are agreeable, the `TState` method could be made `private` to simplify the API.

```diff
diff --git a/osu.Game.Rulesets.Mania/Objects/Drawables/DrawableHoldNote.cs b/osu.Game.Rulesets.Mania/Objects/Drawables/DrawableHoldNote.cs
index 6c70ab3526..2b55e81788 100644
--- a/osu.Game.Rulesets.Mania/Objects/Drawables/DrawableHoldNote.cs
+++ b/osu.Game.Rulesets.Mania/Objects/Drawables/DrawableHoldNote.cs
@@ -265,7 +265,7 @@ protected override void CheckForResult(bool userTriggered, double timeOffset)
             if (Tail.AllJudged)
             {
                 if (Tail.IsHit)
-                    ApplyResult(static (r, _) => r.Type = r.Judgement.MaxResult);
+                    ApplyMaxResult();
                 else
                     MissForcefully();
             }
diff --git a/osu.Game.Rulesets.Mania/Objects/Drawables/DrawableHoldNoteBody.cs b/osu.Game.Rulesets.Mania/Objects/Drawables/DrawableHoldNoteBody.cs
index 731b1b6298..6259033235 100644
--- a/osu.Game.Rulesets.Mania/Objects/Drawables/DrawableHoldNoteBody.cs
+++ b/osu.Game.Rulesets.Mania/Objects/Drawables/DrawableHoldNoteBody.cs
@@ -11,8 +11,6 @@ public partial class DrawableHoldNoteBody : DrawableManiaHitObject<HoldNoteBody>
 
         public override bool DisplayResult => false;
 
-        private bool hit;
-
         public DrawableHoldNoteBody()
             : this(null)
         {
@@ -27,12 +25,10 @@ internal void TriggerResult(bool hit)
         {
             if (AllJudged) return;
 
-            this.hit = hit;
-            ApplyResult(static (r, hitObject) =>
-            {
-                var holdNoteBody = (DrawableHoldNoteBody)hitObject;
-                r.Type = holdNoteBody.hit ? r.Judgement.MaxResult : r.Judgement.MinResult;
-            });
+            if (hit)
+                ApplyMaxResult();
+            else
+                ApplyMinResult();
         }
     }
 }
diff --git a/osu.Game.Rulesets.Mania/Objects/Drawables/DrawableManiaHitObject.cs b/osu.Game.Rulesets.Mania/Objects/Drawables/DrawableManiaHitObject.cs
index 2d10fa27cd..e98622b8bf 100644
--- a/osu.Game.Rulesets.Mania/Objects/Drawables/DrawableManiaHitObject.cs
+++ b/osu.Game.Rulesets.Mania/Objects/Drawables/DrawableManiaHitObject.cs
@@ -87,7 +87,7 @@ protected override void UpdateHitStateTransforms(ArmedState state)
         /// <summary>
         /// Causes this <see cref="DrawableManiaHitObject"/> to get missed, disregarding all conditions in implementations of <see cref="DrawableHitObject.CheckForResult"/>.
         /// </summary>
-        public virtual void MissForcefully() => ApplyResult(static (r, _) => r.Type = r.Judgement.MinResult);
+        public virtual void MissForcefully() => ApplyMinResult();
     }
 
     public abstract partial class DrawableManiaHitObject<TObject> : DrawableManiaHitObject
diff --git a/osu.Game.Rulesets.Mania/Objects/Drawables/DrawableNote.cs b/osu.Game.Rulesets.Mania/Objects/Drawables/DrawableNote.cs
index a70253798a..2246552abe 100644
--- a/osu.Game.Rulesets.Mania/Objects/Drawables/DrawableNote.cs
+++ b/osu.Game.Rulesets.Mania/Objects/Drawables/DrawableNote.cs
@@ -91,12 +91,13 @@ protected override void CheckForResult(bool userTriggered, double timeOffset)
             if (!userTriggered)
             {
                 if (!HitObject.HitWindows.CanBeHit(timeOffset))
-                    ApplyResult(static (r, _) => r.Type = r.Judgement.MinResult);
+                    ApplyMinResult();
 
                 return;
             }
 
             hitResult = HitObject.HitWindows.ResultFor(timeOffset);
+
             if (hitResult == HitResult.None)
                 return;
 
diff --git a/osu.Game.Rulesets.Osu.Tests/TestSceneHitCircle.cs b/osu.Game.Rulesets.Osu.Tests/TestSceneHitCircle.cs
index 8d4145f2c1..abe950f9bb 100644
--- a/osu.Game.Rulesets.Osu.Tests/TestSceneHitCircle.cs
+++ b/osu.Game.Rulesets.Osu.Tests/TestSceneHitCircle.cs
@@ -136,7 +136,7 @@ protected override void CheckForResult(bool userTriggered, double timeOffset)
                 if (auto && !userTriggered && timeOffset > hitOffset && CheckHittable?.Invoke(this, Time.Current, HitResult.Great) == ClickAction.Hit)
                 {
                     // force success
-                    ApplyResult(static (r, _) => r.Type = HitResult.Great);
+                    ApplyResult(HitResult.Great);
                 }
                 else
                     base.CheckForResult(userTriggered, timeOffset);
diff --git a/osu.Game.Rulesets.Osu.Tests/TestSceneHitCircleLateFade.cs b/osu.Game.Rulesets.Osu.Tests/TestSceneHitCircleLateFade.cs
index 2d1e9c1270..838b426cb4 100644
--- a/osu.Game.Rulesets.Osu.Tests/TestSceneHitCircleLateFade.cs
+++ b/osu.Game.Rulesets.Osu.Tests/TestSceneHitCircleLateFade.cs
@@ -208,7 +208,7 @@ protected override void CheckForResult(bool userTriggered, double timeOffset)
                 if (shouldHit && !userTriggered && timeOffset >= 0)
                 {
                     // force success
-                    ApplyResult(static (r, _) => r.Type = HitResult.Great);
+                    ApplyResult(HitResult.Great);
                 }
                 else
                     base.CheckForResult(userTriggered, timeOffset);
diff --git a/osu.Game.Rulesets.Osu/Objects/Drawables/DrawableHitCircle.cs b/osu.Game.Rulesets.Osu/Objects/Drawables/DrawableHitCircle.cs
index ce5422b180..a014ba2e77 100644
--- a/osu.Game.Rulesets.Osu/Objects/Drawables/DrawableHitCircle.cs
+++ b/osu.Game.Rulesets.Osu/Objects/Drawables/DrawableHitCircle.cs
@@ -44,7 +44,6 @@ public partial class DrawableHitCircle : DrawableOsuHitObject, IHasApproachCircl
 
         private Container scaleContainer;
         private InputManager inputManager;
-        private HitResult hitResult;
 
         public DrawableHitCircle()
             : this(null)
@@ -156,12 +155,12 @@ protected override void CheckForResult(bool userTriggered, double timeOffset)
             if (!userTriggered)
             {
                 if (!HitObject.HitWindows.CanBeHit(timeOffset))
-                    ApplyResult(static (r, _) => r.Type = r.Judgement.MinResult);
+                    ApplyMinResult();
 
                 return;
             }
 
-            hitResult = ResultFor(timeOffset);
+            HitResult hitResult = ResultFor(timeOffset);
             var clickAction = CheckHittable?.Invoke(this, Time.Current, hitResult);
 
             if (clickAction == ClickAction.Shake)
@@ -170,20 +169,22 @@ protected override void CheckForResult(bool userTriggered, double timeOffset)
             if (hitResult == HitResult.None || clickAction != ClickAction.Hit)
                 return;
 
-            ApplyResult(static (r, hitObject) =>
+            Vector2? hitPosition = null;
+
+            // Todo: This should also consider misses, but they're a little more interesting to handle, since we don't necessarily know the position at the time of a miss.
+            if (hitResult.IsHit())
             {
-                var hitCircle = (DrawableHitCircle)hitObject;
-                var circleResult = (OsuHitCircleJudgementResult)r;
+                var localMousePosition = ToLocalSpace(inputManager.CurrentState.Mouse.Position);
+                hitPosition = HitObject.StackedPosition + (localMousePosition - DrawSize / 2);
+            }
 
-                // Todo: This should also consider misses, but they're a little more interesting to handle, since we don't necessarily know the position at the time of a miss.
-                if (hitCircle.hitResult.IsHit())
-                {
-                    var localMousePosition = hitCircle.ToLocalSpace(hitCircle.inputManager.CurrentState.Mouse.Position);
-                    circleResult.CursorPositionAtHit = hitCircle.HitObject.StackedPosition + (localMousePosition - hitCircle.DrawSize / 2);
-                }
+            ApplyResult<(HitResult result, Vector2? position)>((r, state) =>
+            {
+                var circleResult = (OsuHitCircleJudgementResult)r;
 
-                circleResult.Type = hitCircle.hitResult;
-            });
+                circleResult.Type = state.result;
+                circleResult.CursorPositionAtHit = state.position;
+            }, (hitResult, hitPosition));
         }
 
         /// <summary>
diff --git a/osu.Game.Rulesets.Osu/Objects/Drawables/DrawableOsuHitObject.cs b/osu.Game.Rulesets.Osu/Objects/Drawables/DrawableOsuHitObject.cs
index 6de60a9d51..5271c03e08 100644
--- a/osu.Game.Rulesets.Osu/Objects/Drawables/DrawableOsuHitObject.cs
+++ b/osu.Game.Rulesets.Osu/Objects/Drawables/DrawableOsuHitObject.cs
@@ -100,12 +100,12 @@ protected override void UpdateInitialTransforms()
         /// <summary>
         /// Causes this <see cref="DrawableOsuHitObject"/> to get hit, disregarding all conditions in implementations of <see cref="DrawableHitObject.CheckForResult"/>.
         /// </summary>
-        public void HitForcefully() => ApplyResult(static (r, _) => r.Type = r.Judgement.MaxResult);
+        public void HitForcefully() => ApplyMaxResult();
 
         /// <summary>
         /// Causes this <see cref="DrawableOsuHitObject"/> to get missed, disregarding all conditions in implementations of <see cref="DrawableHitObject.CheckForResult"/>.
         /// </summary>
-        public void MissForcefully() => ApplyResult(static (r, _) => r.Type = r.Judgement.MinResult);
+        public void MissForcefully() => ApplyMinResult();
 
         private RectangleF parentScreenSpaceRectangle => ((DrawableOsuHitObject)ParentHitObject)?.parentScreenSpaceRectangle ?? Parent!.ScreenSpaceDrawQuad.AABBFloat;
 
diff --git a/osu.Game.Rulesets.Taiko/Objects/Drawables/DrawableDrumRoll.cs b/osu.Game.Rulesets.Taiko/Objects/Drawables/DrawableDrumRoll.cs
index f68198b967..3d7f654fef 100644
--- a/osu.Game.Rulesets.Taiko/Objects/Drawables/DrawableDrumRoll.cs
+++ b/osu.Game.Rulesets.Taiko/Objects/Drawables/DrawableDrumRoll.cs
@@ -143,7 +143,7 @@ protected override void CheckForResult(bool userTriggered, double timeOffset)
             if (timeOffset < 0)
                 return;
 
-            ApplyResult(static (r, _) => r.Type = r.Judgement.MaxResult);
+            ApplyMaxResult();
         }
 
         protected override void UpdateHitStateTransforms(ArmedState state)
diff --git a/osu.Game.Rulesets.Taiko/Objects/Drawables/DrawableDrumRollTick.cs b/osu.Game.Rulesets.Taiko/Objects/Drawables/DrawableDrumRollTick.cs
index aa678d7043..0333fd71a9 100644
--- a/osu.Game.Rulesets.Taiko/Objects/Drawables/DrawableDrumRollTick.cs
+++ b/osu.Game.Rulesets.Taiko/Objects/Drawables/DrawableDrumRollTick.cs
@@ -49,14 +49,14 @@ protected override void CheckForResult(bool userTriggered, double timeOffset)
             if (!userTriggered)
             {
                 if (timeOffset > HitObject.HitWindow)
-                    ApplyResult(static (r, _) => r.Type = r.Judgement.MinResult);
+                    ApplyMinResult();
                 return;
             }
 
             if (Math.Abs(timeOffset) > HitObject.HitWindow)
                 return;
 
-            ApplyResult(static (r, _) => r.Type = r.Judgement.MaxResult);
+            ApplyMaxResult();
         }
 
         public override void OnKilled()
@@ -64,7 +64,7 @@ public override void OnKilled()
             base.OnKilled();
 
             if (Time.Current > HitObject.GetEndTime() && !Judged)
-                ApplyResult(static (r, _) => r.Type = r.Judgement.MinResult);
+                ApplyMinResult();
         }
 
         protected override void UpdateHitStateTransforms(ArmedState state)
diff --git a/osu.Game.Rulesets.Taiko/Objects/Drawables/DrawableFlyingHit.cs b/osu.Game.Rulesets.Taiko/Objects/Drawables/DrawableFlyingHit.cs
index 4349dff9f9..aad9214c5e 100644
--- a/osu.Game.Rulesets.Taiko/Objects/Drawables/DrawableFlyingHit.cs
+++ b/osu.Game.Rulesets.Taiko/Objects/Drawables/DrawableFlyingHit.cs
@@ -25,7 +25,7 @@ public DrawableFlyingHit(DrawableDrumRollTick drumRollTick)
         protected override void LoadComplete()
         {
             base.LoadComplete();
-            ApplyResult(static (r, _) => r.Type = r.Judgement.MaxResult);
+            ApplyMaxResult();
         }
 
         protected override void LoadSamples()
diff --git a/osu.Game.Rulesets.Taiko/Objects/Drawables/DrawableHit.cs b/osu.Game.Rulesets.Taiko/Objects/Drawables/DrawableHit.cs
index cf8e4050ee..ca49ddb7e1 100644
--- a/osu.Game.Rulesets.Taiko/Objects/Drawables/DrawableHit.cs
+++ b/osu.Game.Rulesets.Taiko/Objects/Drawables/DrawableHit.cs
@@ -101,7 +101,7 @@ protected override void CheckForResult(bool userTriggered, double timeOffset)
             if (!userTriggered)
             {
                 if (!HitObject.HitWindows.CanBeHit(timeOffset))
-                    ApplyResult(static (r, _) => r.Type = r.Judgement.MinResult);
+                    ApplyMinResult();
                 return;
             }
 
@@ -110,7 +110,7 @@ protected override void CheckForResult(bool userTriggered, double timeOffset)
                 return;
 
             if (!validActionPressed)
-                ApplyResult(static (r, _) => r.Type = r.Judgement.MinResult);
+                ApplyMinResult();
             else
             {
                 ApplyResult(static (r, hitObject) =>
@@ -217,19 +217,19 @@ protected override void CheckForResult(bool userTriggered, double timeOffset)
 
                 if (!ParentHitObject.Result.IsHit)
                 {
-                    ApplyResult(static (r, _) => r.Type = r.Judgement.MinResult);
+                    ApplyMinResult();
                     return;
                 }
 
                 if (!userTriggered)
                 {
                     if (timeOffset - ParentHitObject.Result.TimeOffset > SECOND_HIT_WINDOW)
-                        ApplyResult(static (r, _) => r.Type = r.Judgement.MinResult);
+                        ApplyMinResult();
                     return;
                 }
 
                 if (Math.Abs(timeOffset - ParentHitObject.Result.TimeOffset) <= SECOND_HIT_WINDOW)
-                    ApplyResult(static (r, _) => r.Type = r.Judgement.MaxResult);
+                    ApplyMaxResult();
             }
 
             public override bool OnPressed(KeyBindingPressEvent<TaikoAction> e)
diff --git a/osu.Game.Rulesets.Taiko/Objects/Drawables/DrawableStrongNestedHit.cs b/osu.Game.Rulesets.Taiko/Objects/Drawables/DrawableStrongNestedHit.cs
index 8f99538448..11759927a9 100644
--- a/osu.Game.Rulesets.Taiko/Objects/Drawables/DrawableStrongNestedHit.cs
+++ b/osu.Game.Rulesets.Taiko/Objects/Drawables/DrawableStrongNestedHit.cs
@@ -27,7 +27,7 @@ public override void OnKilled()
             // it can happen that the hit window of the nested strong hit extends past the lifetime of the parent object.
             // this is a safety to prevent such cases from causing the nested hit to never be judged and as such prevent gameplay from completing.
             if (!Judged && Time.Current > ParentHitObject?.HitObject.GetEndTime())
-                ApplyResult(static (r, _) => r.Type = r.Judgement.MinResult);
+                ApplyMinResult();
         }
     }
 }
diff --git a/osu.Game.Rulesets.Taiko/Objects/Drawables/DrawableSwell.cs b/osu.Game.Rulesets.Taiko/Objects/Drawables/DrawableSwell.cs
index 0781ea5e2a..6eb62cce22 100644
--- a/osu.Game.Rulesets.Taiko/Objects/Drawables/DrawableSwell.cs
+++ b/osu.Game.Rulesets.Taiko/Objects/Drawables/DrawableSwell.cs
@@ -208,7 +208,7 @@ protected override void CheckForResult(bool userTriggered, double timeOffset)
                 expandingRing.ScaleTo(1f + Math.Min(target_ring_scale - 1f, (target_ring_scale - 1f) * completion * 1.3f), 260, Easing.OutQuint);
 
                 if (numHits == HitObject.RequiredHits)
-                    ApplyResult(static (r, _) => r.Type = r.Judgement.MaxResult);
+                    ApplyMaxResult();
             }
             else
             {
diff --git a/osu.Game.Tests/Gameplay/TestSceneDrawableHitObject.cs b/osu.Game.Tests/Gameplay/TestSceneDrawableHitObject.cs
index bf1e52aab5..73177e36e1 100644
--- a/osu.Game.Tests/Gameplay/TestSceneDrawableHitObject.cs
+++ b/osu.Game.Tests/Gameplay/TestSceneDrawableHitObject.cs
@@ -216,7 +216,7 @@ protected override void OnApply()
                     LifetimeStart = LIFETIME_ON_APPLY;
             }
 
-            public void MissForcefully() => ApplyResult(static (r, _) => r.Type = HitResult.Miss);
+            public void MissForcefully() => ApplyResult(HitResult.Miss);
 
             protected override void UpdateHitStateTransforms(ArmedState state)
             {
diff --git a/osu.Game.Tests/Visual/Gameplay/TestScenePoolingRuleset.cs b/osu.Game.Tests/Visual/Gameplay/TestScenePoolingRuleset.cs
index 00bd58e303..b567e8de8d 100644
--- a/osu.Game.Tests/Visual/Gameplay/TestScenePoolingRuleset.cs
+++ b/osu.Game.Tests/Visual/Gameplay/TestScenePoolingRuleset.cs
@@ -431,7 +431,7 @@ protected override void OnApply()
             protected override void CheckForResult(bool userTriggered, double timeOffset)
             {
                 if (timeOffset > HitObject.Duration)
-                    ApplyResult(static (r, _) => r.Type = r.Judgement.MaxResult);
+                    ApplyMaxResult();
             }
 
             protected override void UpdateHitStateTransforms(ArmedState state)
@@ -468,7 +468,7 @@ protected override void UpdateHitStateTransforms(ArmedState state)
             public override void OnKilled()
             {
                 base.OnKilled();
-                ApplyResult(static (r, _) => r.Type = r.Judgement.MinResult);
+                ApplyMinResult();
             }
         }
 
@@ -547,7 +547,7 @@ protected override void CheckForResult(bool userTriggered, double timeOffset)
             {
                 base.CheckForResult(userTriggered, timeOffset);
                 if (timeOffset >= 0)
-                    ApplyResult(static (r, _) => r.Type = r.Judgement.MaxResult);
+                    ApplyMaxResult();
             }
         }
 
@@ -596,7 +596,7 @@ protected override void CheckForResult(bool userTriggered, double timeOffset)
             {
                 base.CheckForResult(userTriggered, timeOffset);
                 if (timeOffset >= 0)
-                    ApplyResult(static (r, _) => r.Type = r.Judgement.MaxResult);
+                    ApplyMaxResult();
             }
         }
 
diff --git a/osu.Game/Rulesets/Judgements/JudgementResult.cs b/osu.Game/Rulesets/Judgements/JudgementResult.cs
index b781a13929..4b98df50d7 100644
--- a/osu.Game/Rulesets/Judgements/JudgementResult.cs
+++ b/osu.Game/Rulesets/Judgements/JudgementResult.cs
@@ -30,7 +30,7 @@ public class JudgementResult
 
         /// <summary>
         /// The time at which this <see cref="JudgementResult"/> occurred.
-        /// Populated when this <see cref="JudgementResult"/> is applied via <see cref="DrawableHitObject.ApplyResult"/>.
+        /// Populated when this <see cref="JudgementResult"/> is applied via <see cref="DrawableHitObject.ApplyResult{T}"/>.
         /// </summary>
         /// <remarks>
         /// This is used instead of <see cref="TimeAbsolute"/> to check whether this <see cref="JudgementResult"/> should be reverted.
diff --git a/osu.Game/Rulesets/Objects/Drawables/DrawableHitObject.cs b/osu.Game/Rulesets/Objects/Drawables/DrawableHitObject.cs
index e30ce13f08..07fab72814 100644
--- a/osu.Game/Rulesets/Objects/Drawables/DrawableHitObject.cs
+++ b/osu.Game/Rulesets/Objects/Drawables/DrawableHitObject.cs
@@ -682,17 +682,28 @@ public virtual void OnKilled()
             UpdateResult(false);
         }
 
+        protected void ApplyMaxResult() => ApplyResult((r, _) => r.Type = r.Judgement.MaxResult);
+        protected void ApplyMinResult() => ApplyResult((r, _) => r.Type = r.Judgement.MinResult);
+
+        protected void ApplyResult(HitResult type) => ApplyResult(static (result, state) => result.Type = state, type);
+
+        [Obsolete("Use overload with state, preferrably with static delegates to avoid allocation overhead.")] // Can be removed 2024-07-26
+        protected void ApplyResult(Action<JudgementResult> application) => ApplyResult((r, _) => application(r), this);
+
+        protected void ApplyResult(Action<JudgementResult, DrawableHitObject> application) => ApplyResult(application, this);
+
         /// <summary>
         /// Applies the <see cref="Result"/> of this <see cref="DrawableHitObject"/>, notifying responders such as
         /// the <see cref="ScoreProcessor"/> of the <see cref="JudgementResult"/>.
         /// </summary>
         /// <param name="application">The callback that applies changes to the <see cref="JudgementResult"/>. Using a `static` delegate is recommended to avoid allocation overhead.</param>
-        protected void ApplyResult(Action<JudgementResult, DrawableHitObject> application)
+        /// <param name="state">The state.</param>
+        protected void ApplyResult<T>(Action<JudgementResult, T> application, T state)
         {
             if (Result.HasResult)
                 throw new InvalidOperationException("Cannot apply result on a hitobject that already has a result.");
 
-            application?.Invoke(Result, this);
+            application?.Invoke(Result, state);
 
             if (!Result.HasResult)
                 throw new InvalidOperationException($"{GetType().ReadableName()} applied a {nameof(JudgementResult)} but did not update {nameof(JudgementResult.Type)}.");
@@ -737,7 +748,7 @@ protected bool UpdateResult(bool userTriggered)
         /// Checks if a scoring result has occurred for this <see cref="DrawableHitObject"/>.
         /// </summary>
         /// <remarks>
-        /// If a scoring result has occurred, this method must invoke <see cref="ApplyResult"/> to update the result and notify responders.
+        /// If a scoring result has occurred, this method must invoke <see cref="ApplyResult{T}"/> to update the result and notify responders.
         /// </remarks>
         /// <param name="userTriggered">Whether the user triggered this check.</param>
         /// <param name="timeOffset">The offset from the end time of the <see cref="HitObject"/> at which this check occurred.

```

Reply:
Chiming in to give my two cents. I think creating fields that are only used within `ApplyResult` is unnecessary bloat to the class. If you can pass the data as a parameter and avoid creating a single use field for it, I think passing it as a parameter under `TState` is the better alternative.

Reply:
The primary reason why I was against the generic in the first place that I have seen @peppy express distaste towards generics and the havoc they tend to wreak on things once everything becomes overly specialised. If you're fine with the generic, then so am I, but then again, the _only_ usage that is slightly teeth-pain-inducing here is the `DrawableHitCircle` one, and even then, you could - as I suggested before - call `ResultFor()` twice to avoid the field.

For every other instance in this diff, such as

https://github.com/ppy/osu/blob/6cfd2813ede27126f002ace78e3def0e5f7b03f5/osu.Game.Rulesets.Mania/Objects/Drawables/DrawableHoldNoteBody.cs#L31-L35

any extra added field can be trivially eliminated by hoisting the branch _out_ of the delegate, as per patch above.

Reply:
I've applied the above patch with follow-up removals of some other leftover weirdness.